File: data-oriented/01-unified-vs-disjoint/java/Main.java
import java.util.ArrayList;
import java.util.Scanner;
import java.io.BufferedReader;
import java.io.InputStreamReader;

interface Num {
    float toFloat();
}

class NumInt implements Num {
    int val;
    public NumInt(int v) {
        val = v;
    }

    public float toFloat() {
        return (float)val;
    }
}

class NumFloat implements Num {
    float val;
    public NumFloat(float v) {
        val = v;
    }

    public float toFloat() {
        return val;
    }
}

public class Main {
    static int ITERS = 16;

    public static void main(String[] args) {
        if (args[0].equals("oo")) {
            ooWay();
        } else if (args[0].equals("do")) {
            doWay();
        } else if (args[0].equals("array")) {
            arrayWay();
        }
    }

    static void ooWay() {
        ArrayList<Num> floats = new ArrayList<>();
        BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
        long top = System.nanoTime();

        try {
            String line = stdin.readLine();
            while (line != null) {
                try {
                    floats.add(new NumInt(Integer.parseInt(line)));
                } catch (Exception e1) {
                    try {
                        floats.add(new NumFloat(Float.parseFloat(line)));
                    } catch (Exception e2) {
                    }
                }
                line = stdin.readLine();
            }
        } catch (Exception e3) {
        }

        System.err.println("read time: " + humanReadableTime(System.nanoTime() - top));

        for (int t = 0; t < ITERS; ++t) {
            top = System.nanoTime();
            float sum = 0;
            for (Num f: floats) {
                sum += f.toFloat();
            }
            System.out.println("sum: " + sum);
            System.err.println("sum time: " + humanReadableTime(System.nanoTime() - top));
        }
    }

    static void doWay() {
        ArrayList<Integer> ints = new ArrayList<>();
        ArrayList<Float> floats = new ArrayList<>();

        BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
        long top = System.nanoTime();

        try {
            String line = stdin.readLine();
            while (line != null) {
                try {
                    ints.add(Integer.parseInt(line));
                } catch (Exception e1) {
                    try {
                        floats.add(Float.parseFloat(line));
                    } catch (Exception e2) {
                    }
                }
                line = stdin.readLine();
            }
        } catch (Exception e3) {
        }
        System.err.println("read time: " + humanReadableTime(System.nanoTime() - top));

        for (int t = 0; t < ITERS; ++t) {
            top = System.nanoTime();
            float sum = 0;
            for (int i: ints) {
                sum += i;
            }
            for (float f: floats) {
                sum += f;
            }
            System.out.println("sum: " + sum);
            System.err.println("sum time: " + humanReadableTime(System.nanoTime() - top));
        }
    }

    static void arrayWay() {
        int[] ints = new int[2000000];
        float[] floats = new float[2000000];
        int i = 0, j = 0;

        BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
        long top = System.nanoTime();
        try {
            String line = stdin.readLine();
            while (line != null) {
                try {
                    ints[i++] = Integer.parseInt(line);
                } catch (Exception e1) {
                    try {
                        floats[j++] = Float.parseFloat(line);
                    } catch (Exception e2) {
                    }
                }
                line = stdin.readLine();
            }
        } catch (Exception e3) {
        }
        System.err.println("read time: " + humanReadableTime(System.nanoTime() - top));

        for (int t = 0; t < ITERS; ++t) {
            top = System.nanoTime();
            float sum = 0;
            for (int x: ints) {
                sum += x;
            }
            for (float f: floats) {
                sum += f;
            }
            System.out.println("sum: " + sum);
            System.err.println("sum time: " + humanReadableTime(System.nanoTime() - top));
        }
    }

    static String humanReadableTime(long nano) {
        if (nano < 1000) {
            return nano + " ns";
        } else if (nano < 1000000) {
            double micro = nano / 1000.0;
            return micro + " us";
        } else if (nano < 1000000000) {
            double milli = nano / 1000000.0;
            return milli + " ms";
        } else {
            double sec = nano / 1000000000.0;
            return sec + " s";
        }
    }
}

File: data-oriented/01-unified-vs-disjoint/rust/main.rs
use std::env::args;
use std::error::Error;
use std::io::{self, BufRead};
use std::time::Instant;

const ITERS: usize = 16;

enum Num {
    Int(i32),
    Float(f32),
}

trait NumTrait {
    fn to_float(&self) -> f32;
}

impl NumTrait for i32 {
    fn to_float(&self) -> f32 {
        *self as f32
    }
}

impl NumTrait for f32 {
    fn to_float(&self) -> f32 {
        *self
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = args().collect();
    if args[1] == "enum" {
        do_enum()?;
    } else if args[1] == "do" {
        do_do()?;
    } else if args[1] == "trait" {
        do_trait()?;
    }
    return Ok(());
}

fn do_enum() -> Result<(), Box<dyn Error>> {
    let stdin = io::stdin();
    let stdin = stdin.lock();
    let mut v: Vec<Num> = Vec::with_capacity(4096);

    let top = Instant::now();
    for line in stdin.lines() {
        let line = line?;
        match line.parse::<i32>() {
            Ok(i) => v.push(Num::Int(i)),
            Err(_) => {
                match line.parse::<f32>() {
                    Ok(f) => v.push(Num::Float(f)),
                    Err(_) => { }
                }
            }
        }
    }
    eprintln!("read time: {:?}", top.elapsed());

    for _ in 0 .. ITERS {
        let top = Instant::now();
        let mut sum: f32 = 0.0;
        for n in &v {
            match *n {
                Num::Int(i) => sum += i as f32,
                Num::Float(f) => sum += f,
            }
        }
        println!("sum: {}", sum);
        eprintln!("sum time: {:?}", top.elapsed());
    }

    return Ok(());
}

fn do_do() -> Result<(), Box<dyn Error>> {
    let stdin = io::stdin();
    let stdin = stdin.lock();
    let mut ints: Vec<i32> = Vec::with_capacity(4096);
    let mut floats: Vec<f32> = Vec::with_capacity(4096);

    let top = Instant::now();
    for line in stdin.lines() {
        let line = line?;
        match line.parse::<i32>() {
            Ok(i) => ints.push(i),
            Err(_) => {
                match line.parse::<f32>() {
                    Ok(f) => floats.push(f),
                    Err(_) => { }
                }
            }
        }
    }
    eprintln!("read time: {:?}", top.elapsed());

    for _ in 0 .. ITERS {
        let top = Instant::now();
        let mut sum: f32 = 0.0;
        for n in &ints { sum += *n as f32; }
        for n in &floats { sum += *n; }
        println!("sum: {}", sum);
        eprintln!("sum time: {:?}", top.elapsed());
    }

    return Ok(());
}

fn do_trait() -> Result<(), Box<dyn Error>> {
    let stdin = io::stdin();
    let stdin = stdin.lock();
    let mut v: Vec<Box<dyn NumTrait>> = Vec::with_capacity(4096);

    let top = Instant::now();
    for line in stdin.lines() {
        let line = line?;
        match line.parse::<i32>() {
            Ok(i) => v.push(Box::new(i)),
            Err(_) => {
                match line.parse::<f32>() {
                    Ok(f) => v.push(Box::new(f)),
                    Err(_) => { }
                }
            }
        }
    }
    eprintln!("read time: {:?}", top.elapsed());

    for _ in 0 .. ITERS {
        let top = Instant::now();
        let mut sum: f32 = 0.0;
        for n in &v {
            sum += n.to_float();
        }
        println!("sum: {}", sum);
        eprintln!("sum time: {:?}", top.elapsed());
    }

    return Ok(());
}
